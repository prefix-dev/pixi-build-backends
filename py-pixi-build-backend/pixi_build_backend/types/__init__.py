# Re-export all the types
from .metadata_provider import MetadataProvider as MetadataProvider