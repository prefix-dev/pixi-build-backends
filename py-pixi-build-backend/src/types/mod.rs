mod config;
mod generated_recipe;
mod platform;
mod project_model;
mod python_params;

pub use generated_recipe::{PyGenerateRecipe, PyGeneratedRecipe, PyVecString};
pub use platform::PyPlatform;
pub use project_model::PyProjectModelV1;

pub use config::PyBackendConfig;
pub use python_params::PyPythonParams;
