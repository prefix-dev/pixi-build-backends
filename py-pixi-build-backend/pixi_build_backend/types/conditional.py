from typing import Optional
from typing_extensions import TypeAlias
from pixi_build_backend.pixi_build_backend import (
    PyConditionalString,
    PyListOrItemString,
    PyListOrItemPackageDependency,
    PyConditionalPackageDependency,
)


# Type aliases for FFI types
ListOrItemString: TypeAlias = PyListOrItemString
ListOrItemPackageDependency: TypeAlias = PyListOrItemPackageDependency


class ConditionalString:
    _inner: PyConditionalString

    def __init__(self, condition: str, then: ListOrItemString, else_: Optional[ListOrItemString]) -> None:
        else_ = else_ if else_ is not None else ListOrItemString([])
        self._inner = PyConditionalString(condition, then, else_)

    @property
    def condition(self) -> str:
        """Get the condition string."""
        return self._inner.condition

    @property
    def then_value(self) -> "ListOrItemString":
        """Get the then value."""
        return self._inner.then_value
        # return result

    @property
    def else_value(self) -> ListOrItemString:
        """Get the else value."""
        return self._inner.else_value

    def __str__(self) -> str:
        return str(self._inner)

    def __eq__(self, other: object) -> bool:
        """Check equality."""
        if not isinstance(other, ConditionalString):
            return False
        return self._inner == other._inner


class ConditionalPackageDependency:
    _inner: PyConditionalPackageDependency

    def __init__(self, condition: str, then: ListOrItemPackageDependency, else_: ListOrItemPackageDependency) -> None:
        self._inner = PyConditionalPackageDependency(condition, then, else_)

    @property
    def condition(self) -> str:
        """Get the condition string."""
        return self._inner.condition

    @condition.setter
    def condition(self, value: str) -> None:
        """Set the condition string."""
        self._inner.condition = value

    @property
    def then_value(self) -> ListOrItemPackageDependency:
        """Get the then value."""
        return self._inner.then_value

    @property
    def else_value(self) -> ListOrItemPackageDependency:
        """Get the else value."""
        return self._inner.else_value
