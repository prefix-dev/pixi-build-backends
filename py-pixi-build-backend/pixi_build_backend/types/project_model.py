from typing import Optional, List
from pixi_build_backend.pixi_build_backend import (
    PyProjectModelV1,
)


class ProjectModelV1:
    """A project model version 1."""

    _inner: PyProjectModelV1

    def __init__(self, name: str, version: Optional[str] = None):
        self._inner = PyProjectModelV1(name, version)

    @property
    def name(self) -> str:
        """
        Get the project name.

        Examples
        --------
        ```python
        >>> model = ProjectModelV1("my-project")
        >>> model.name
        'my-project'
        >>>
        ```
        """
        return self._inner.name

    @classmethod
    def _from_py(cls, model: PyProjectModelV1) -> "ProjectModelV1":
        """Create a ProjectModelV1 from a FFI PyProjectModelV1."""
        instance = cls.__new__(cls)
        instance._inner = model
        return instance

    @property
    def version(self) -> Optional[str]:
        """
        Get the project version.

        Examples
        --------
        ```python
        >>> model = ProjectModelV1("my-project", "1.0.0")
        >>> model.version
        '1.0.0'
        >>> ProjectModelV1("test").version is None
        True
        >>>
        ```
        """
        return self._inner.version

    @property
    def description(self) -> Optional[str]:
        """
        Get the project description.

        Examples
        --------
        ```python
        >>> model = ProjectModelV1("test")
        >>> model.description is None
        True
        >>>
        ```
        """
        return self._inner.description

    @property
    def authors(self) -> Optional[List[str]]:
        """Get the project authors."""
        return self._inner.authors

    @property
    def license(self) -> Optional[str]:
        """Get the project license."""
        return self._inner.license

    @property
    def license_file(self) -> Optional[str]:
        """Get the project license file path."""
        return self._inner.license_file

    @property
    def readme(self) -> Optional[str]:
        """Get the project readme file path."""
        return self._inner.readme

    @property
    def homepage(self) -> Optional[str]:
        """
        Get the project homepage URL.

        Examples
        --------
        ```python
        >>> model = ProjectModelV1("test")
        >>> model.homepage is None
        True
        >>>
        ```
        """
        return self._inner.homepage

    @property
    def repository(self) -> Optional[str]:
        """Get the project repository URL."""
        return self._inner.repository

    @property
    def documentation(self) -> Optional[str]:
        """Get the project documentation URL."""
        return self._inner.documentation

    def __repr__(self) -> str:
        return self._inner.__repr__()

    def _debug_str(self) -> str:
        """Get a debug string representation of the project model."""
        return self._inner._debug_str()
