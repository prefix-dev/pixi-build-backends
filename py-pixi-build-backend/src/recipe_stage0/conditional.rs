use pyo3::exceptions::PyTypeError;
use pyo3::types::PyAnyMethods;
use pyo3::{Bound, FromPyObject, PyAny, PyErr, intern, pyclass, pymethods};
use recipe_stage0::matchspec::PackageDependency;
use recipe_stage0::recipe::{Conditional, Item, ListOrItem};

macro_rules! create_py_item {
    ($name: ident, $type: ident) => {
        #[pyclass]
        #[derive(Clone)]
        pub struct $name {
            pub(crate) inner: Item<$type>,
        }

        #[pymethods]
        impl $name {
            pub fn is_value(&self) -> bool {
                matches!(self.inner, Item::Value(_))
            }

            pub fn is_conditional(&self) -> bool {
                matches!(self.inner, Item::Conditional(_))
            }
        }
    };
}

create_py_item!(PyItemPackageDependency, PackageDependency);
create_py_item!(PyItemString, String);

macro_rules! create_pylist_or_item {
    ($name: ident, $type: ident) => {
        #[pyclass]
        #[derive(Clone)]
        pub struct $name {
            pub(crate) inner: ListOrItem<$type>,
        }

        #[pymethods]
        impl $name {
            #[staticmethod]
            pub fn single(item: $type) -> Self {
                $name {
                    inner: ListOrItem::single(item),
                }
            }

            #[staticmethod]
            pub fn list(items: Vec<$type>) -> Self {
                $name {
                    inner: ListOrItem::new(items),
                }
            }

            pub fn is_single(&self) -> bool {
                self.inner.len() == 1
            }

            pub fn is_list(&self) -> bool {
                self.inner.len() > 1
            }

            pub fn get_single(&self) -> Option<String> {
                if self.inner.len() == 1 {
                    self.inner.0.first().cloned()
                } else {
                    None
                }
            }

            pub fn get_list(&self) -> Vec<String> {
                self.inner.0.clone()
            }
        }
    };
}

create_pylist_or_item!(PyListOrItemString, String);

macro_rules! create_conditional_interface {
    ($name: ident, $type: ident) => {
        paste::paste! {
            #[pyclass]
            #[derive(Clone)]
            pub struct $name {
                pub(crate) inner: Conditional<$type>,
            }

            #[pymethods]
            impl $name {
                #[new]
                pub fn new(
                    condition: String,
                    then_value: [<PyListOrItem $type>],
                    else_value: Option<[<PyListOrItem $type>]>,
                ) -> Self {
                    $name {
                        inner: Conditional {
                            condition,
                            then: then_value.inner,
                            else_value: else_value.map(|e| e.inner).unwrap_or_default(),
                        },
                    }
                }

                #[getter]
                pub fn condition(&self) -> $type {
                    self.inner.condition.clone()
                }

                #[getter]
                pub fn then_value(&self) -> [<PyListOrItem $type>] {
                    [<PyListOrItem $type>] {
                        inner: self.inner.then.clone(),
                    }
                }

                #[getter]
                pub fn else_value(&self) -> [<PyListOrItem $type>] {
                    [<PyListOrItem $type>] {
                        inner: self.inner.else_value.clone(),
                    }
                }
            }
        }
    };
}

create_conditional_interface!(PyConditionalString, String);

impl<'a> TryFrom<Bound<'a, PyAny>> for PyItemPackageDependency {
    type Error = PyErr;
    fn try_from(value: Bound<'a, PyAny>) -> Result<Self, Self::Error> {
        let intern_val = intern!(value.py(), "_inner");
        if !value.hasattr(intern_val)? {
            return Err(PyTypeError::new_err(
                "object is not a PackageDependency type",
            ));
        }

        let inner = value.getattr(intern_val)?;
        if !inner.is_instance_of::<Self>() {
            return Err(PyTypeError::new_err("'_inner' is invalid"));
        }

        PyItemPackageDependency::extract_bound(&inner)
    }
}

impl From<Conditional<String>> for PyConditionalString {
    fn from(conditional: Conditional<String>) -> Self {
        PyConditionalString { inner: conditional }
    }
}

impl From<PyConditionalString> for Conditional<String> {
    fn from(py_conditional: PyConditionalString) -> Self {
        py_conditional.inner
    }
}

impl From<ListOrItem<String>> for PyListOrItemString {
    fn from(list_or_item: ListOrItem<String>) -> Self {
        PyListOrItemString {
            inner: list_or_item,
        }
    }
}

impl From<PyListOrItemString> for ListOrItem<String> {
    fn from(py_list_or_item: PyListOrItemString) -> Self {
        py_list_or_item.inner
    }
}
