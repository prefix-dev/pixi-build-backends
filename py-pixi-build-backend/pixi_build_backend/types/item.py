from typing import List, Optional
from pixi_build_backend.pixi_build_backend import PyConditionalString, PyListOrItemString, PyVecItemPackageDependency, PyItemPackageDependency
from pixi_build_backend.types.conditional import ConditionalPackageDependency
from pixi_build_backend.types.requirements import PackageDependency

class VecItemPackageDependency(List):
    """A wrapper for a list of ItemPackageDependency."""

    _inner: PyVecItemPackageDependency

    def __init__(self, items=None):
        """Initialize with optional items."""
        if items is None:
            self._inner = PyVecItemPackageDependency()
        else:
            # Extract _inner from ItemPackageDependency objects
            inner_items = [item._inner if isinstance(item, ItemPackageDependency) else item for item in items]
            self._inner = PyVecItemPackageDependency(inner_items)

    @classmethod
    def _from_inner(cls, inner: PyVecItemPackageDependency) -> "VecItemPackageDependency":
        """Create from PyVecItemPackageDependency."""
        instance = cls.__new__(cls)
        instance._inner = inner
        return instance

    def __len__(self) -> int:
        """Return the length of the list."""
        return len(self._inner)

    def __getitem__(self, index) -> "ItemPackageDependency":
        """Get item at index."""
        return ItemPackageDependency._from_inner(self._inner[index])

    def __setitem__(self, index, value):
        """Set item at index."""
        inner_value = value._inner if isinstance(value, ItemPackageDependency) else value
        self._inner[index] = inner_value

    def __delitem__(self, index):
        """Delete item at index."""
        del self._inner[index]

    def __iter__(self):
        """Return iterator."""
        for item in self._inner:
            yield ItemPackageDependency._from_inner(item)

    def __contains__(self, item: "ItemPackageDependency") -> bool:
        """Check if item is in the list."""
        inner_item = item._inner if isinstance(item, ItemPackageDependency) else item
        return inner_item in self._inner

    def append(self, item: "ItemPackageDependency"):
        """Append item to the list."""
        inner_item = item._inner
        self._inner.append(inner_item)

    def extend(self, items: List["ItemPackageDependency"]):
        """Extend the list with items."""
        inner_items = [item._inner for item in items]
        self._inner.extend(inner_items)

    def insert(self, index, item: "ItemPackageDependency"):
        """Insert item at index."""
        inner_item = item._inner
        self._inner.insert(index, inner_item)

    def remove(self, item: "ItemPackageDependency"):
        """Remove first occurrence of item."""
        inner_item = item._inner
        self._inner.remove(inner_item)

    def pop(self, index=-1):
        """Remove and return item at index."""
        return ItemPackageDependency._from_inner(self._inner.pop(index))

    def clear(self):
        """Remove all items."""
        self._inner.clear()

    def index(self, item: "ItemPackageDependency", start=0, stop=None):
        """Return index of first occurrence of item."""
        inner_item = item._inner
        if stop is None:
            return self._inner.index(inner_item, start)
        else:
            return self._inner.index(inner_item, start, stop)

    def count(self, item) -> int:
        """Return count of occurrences of item."""
        inner_item = item._inner if isinstance(item, "ItemPackageDependency") else item
        return self._inner.count(inner_item)

    def reverse(self):
        """Reverse the list in place."""
        self._inner.reverse()

    def sort(self, key=None, reverse=False):
        """Sort the list in place."""
        if key is None:
            self._inner.sort(reverse=reverse)
        else:
            self._inner.sort(key=key, reverse=reverse)

    def copy(self):
        """Return a shallow copy."""
        return VecItemPackageDependency._from_inner(self._inner.copy())

    def __eq__(self, other) -> bool:
        """Check equality."""
        if isinstance(other, VecItemPackageDependency):
            return self._inner == other._inner
        return False

    def __str__(self) -> str:
        """Return string representation."""
        return str(self._inner)
    

class ItemPackageDependency:
    """A package dependency item wrapper."""

    _inner: PyItemPackageDependency

    def __init__(self, name: str):
        self._inner = PyItemPackageDependency(name)

    @classmethod
    def new_from_conditional(
        cls,
        conditional: ConditionalPackageDependency
    ) -> "ItemPackageDependency":
        new_class = cls.__new__(cls)
        new_class._inner = PyItemPackageDependency.new_from_conditional(conditional._inner)
        return new_class


    @classmethod
    def _from_inner(cls, inner: PyItemPackageDependency) -> "ItemPackageDependency":
        """Create an ItemPackageDependency from a FFI PyItemPackageDependency."""
        instance = cls.__new__(cls)
        instance._inner = inner
        return instance

    def __str__(self):
        return str(self._inner)
    
    @property
    def concrete(self) -> "PackageDependency":
        """Get the concrete package dependency."""
        return PackageDependency._from_inner(self._inner.concrete())
    
    @property
    def template(self) -> Optional[str]:
        """Get the template string if this is a template."""
        return self._inner.template()
    
    @property
    def conditional(self) -> Optional[ConditionalPackageDependency]:
        """Get the conditional string if this is a conditional."""
        return self._inner.conditional()