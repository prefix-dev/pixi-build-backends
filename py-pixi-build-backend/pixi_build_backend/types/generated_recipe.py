from typing import Any, Dict, Optional, Protocol, List
from pathlib import Path
from pixi_build_backend.types.intermediate_recipe import IntermediateRecipe
from pixi_build_backend.pixi_build_backend import PyGeneratedRecipe, PyGenerateRecipe
from pixi_build_backend.types.platform import Platform
from pixi_build_backend.types.project_model import ProjectModelV1
from pixi_build_backend.types.python_params import PythonParams


class GeneratedRecipe:
    """A generated recipe wrapper."""

    _inner: PyGeneratedRecipe

    def __init__(self) -> None:
        self._inner = PyGeneratedRecipe()

    @classmethod
    def from_model(cls, model: ProjectModelV1, manifest_root: Path) -> "GeneratedRecipe":
        """Create a GeneratedRecipe from a ProjectModelV1."""
        instance = cls()
        instance._inner = PyGeneratedRecipe().from_model(model._inner, manifest_root)
        return instance

    @property
    def recipe(self) -> IntermediateRecipe:
        """Get the recipe."""
        return IntermediateRecipe._from_inner(self._inner.recipe)

    @property
    def metadata_input_globs(self) -> List[str]:
        """Get the metadata input globs."""
        return self._inner.metadata_input_globs

    @property
    def build_input_globs(self) -> List[str]:
        """Get the build input globs."""
        return self._inner.build_input_globs

    def __repr__(self) -> str:
        return self._inner.__repr__()

    def _into_py(self) -> PyGeneratedRecipe:
        """
        Converts this object into a type that can be used by the Rust code.
        """
        return self._inner


class GenerateRecipeProtocol(Protocol):
    """
    Protocol for generating recipes.
    This should be implemented by the Python generator.
    """

    def generate_recipe(
        self,
        model: ProjectModelV1,
        config: Dict[str, Any],
        manifest_path: str,
        host_platform: Platform,
        python_params: Optional[PythonParams] = None,
    ) -> GeneratedRecipe:
        """Generate a recipe for a Python package."""
        ...


class GenerateRecipe:
    """Protocol for generating recipes."""

    _inner: PyGenerateRecipe

    def __init__(self, instance: GenerateRecipeProtocol):
        self._inner = PyGenerateRecipe(instance)

    def _into_py(self) -> PyGenerateRecipe:
        """
        Converts this object into a type that can be used by the Rust code.
        """
        return self._inner
