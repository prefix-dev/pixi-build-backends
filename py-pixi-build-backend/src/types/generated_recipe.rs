use std::path::PathBuf;

use miette::IntoDiagnostic;
use pixi_build_backend::generated_recipe::{GenerateRecipe, GeneratedRecipe};
use pyo3::{
    PyObject, Python, pyclass, pymethods,
    types::{PyAnyMethods, PyString},
};

use crate::{
    recipe_stage0::recipe::PyIntermediateRecipe,
    types::{PyBackendConfig, PyPlatform, PyProjectModelV1, PyPythonParams},
};

#[pyclass]
#[derive(Clone, Default)]
pub struct PyGeneratedRecipe {
    pub(crate) inner: pixi_build_backend::generated_recipe::GeneratedRecipe,
}

#[pymethods]
impl PyGeneratedRecipe {
    #[new]
    pub fn new() -> Self {
        PyGeneratedRecipe {
            inner: pixi_build_backend::generated_recipe::GeneratedRecipe::default(),
        }
    }

    #[staticmethod]
    pub fn from_model(model: PyProjectModelV1, manifest_root: PathBuf) -> Self {
        let recipe = GeneratedRecipe::from_model(model.inner.clone(), manifest_root);
        PyGeneratedRecipe { inner: recipe }
    }

    #[getter]
    pub fn recipe(&self) -> PyIntermediateRecipe {
        self.inner.recipe.clone().into()
    }

    #[getter]
    pub fn metadata_input_globs(&self) -> Vec<String> {
        self.inner.metadata_input_globs.iter().cloned().collect()
    }

    #[getter]
    pub fn build_input_globs(&self) -> Vec<String> {
        self.inner.build_input_globs.iter().cloned().collect()
    }
}

impl From<GeneratedRecipe> for PyGeneratedRecipe {
    fn from(recipe: GeneratedRecipe) -> Self {
        PyGeneratedRecipe { inner: recipe }
    }
}

impl From<PyGeneratedRecipe> for GeneratedRecipe {
    fn from(py_recipe: PyGeneratedRecipe) -> Self {
        py_recipe.inner
    }
}

/// Trait part
#[pyclass]
#[derive(Clone)]
pub struct PyGenerateRecipe {
    model: PyObject,
}

#[pymethods]
impl PyGenerateRecipe {
    #[new]
    pub fn new(model: PyObject) -> Self {
        PyGenerateRecipe { model }
    }
}

impl GenerateRecipe for PyGenerateRecipe {
    type Config = PyBackendConfig;

    fn generate_recipe(
        &self,
        model: &pixi_build_types::ProjectModelV1,
        config: &Self::Config,
        manifest_path: std::path::PathBuf,
        host_platform: rattler_conda_types::Platform,
        python_params: Option<pixi_build_backend::generated_recipe::PythonParams>,
    ) -> miette::Result<pixi_build_backend::generated_recipe::GeneratedRecipe> {
        let recipe: GeneratedRecipe = Python::with_gil(|py| {
            let manifest_str = manifest_path.to_string_lossy().to_string();

            // we dont pass the wrapper but the python inner model directly
            let py_object = config.model.clone();

            // For other types, we try to wrap them into the Python class
            // So user can use the Python API
            let project_model_class = py
                .import("pixi_build_backend.types.project_model")
                .into_diagnostic()?
                .getattr("ProjectModelV1")
                .into_diagnostic()?;

            let project_model = project_model_class
                .call_method1("_from_py", (PyProjectModelV1::from(model),))
                .into_diagnostic()?;

            let platform_model_class = py
                .import("pixi_build_backend.types.platform")
                .into_diagnostic()?
                .getattr("Platform")
                .into_diagnostic()?;

            let platform_model = platform_model_class
                .call_method1("_from_py", (PyPlatform::from(host_platform),))
                .into_diagnostic()?;

            let python_params_class = py
                .import("pixi_build_backend.types.python_params")
                .into_diagnostic()?
                .getattr("PythonParams")
                .into_diagnostic()?;
            let python_params_model = python_params_class
                .call_method1(
                    "_from_py",
                    (PyPythonParams::from(python_params.unwrap_or_default()),),
                )
                .into_diagnostic()?;

            let generated_recipe_py = self
                .model
                .bind(py)
                .call_method(
                    "generate_recipe",
                    (
                        project_model,
                        py_object,
                        PyString::new(py, manifest_str.as_str()),
                        platform_model,
                        python_params_model,
                    ),
                    None,
                )
                .into_diagnostic()?;

            // To expose a nice API for the user, we extract the PyGeneratedRecipe
            // calling private _into_py method
            let generated_recipe: PyGeneratedRecipe = generated_recipe_py
                .call_method0("_into_py")
                .into_diagnostic()?
                .extract::<PyGeneratedRecipe>()
                .into_diagnostic()?;

            Ok::<_, miette::Report>(generated_recipe.into())
        })?;

        Ok(recipe)
    }
}
