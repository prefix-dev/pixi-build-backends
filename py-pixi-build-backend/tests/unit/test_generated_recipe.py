from pathlib import Path
from typing import Any
from pixi_build_backend.types.conditional import ConditionalPackageDependency, ListOrItemPackageDependency
from pixi_build_backend.types.generated_recipe import GeneratedRecipe
from pixi_build_backend.types.item import ItemPackageDependency
from pixi_build_backend.types.project_model import ProjectModelV1


def test_generated_recipe_from_model(snapshot: Any) -> None:
    """Test initialization of ProjectModelV1."""
    model = ProjectModelV1(name="test_project", version="1.0.0")

    generated_recipe = GeneratedRecipe.from_model(model, Path("."))

    print(type(generated_recipe.recipe))

    assert snapshot == generated_recipe.recipe.to_yaml()


def test_setting_package_name_from_generated_recipe() -> None:
    """Test initialization of ProjectModelV1."""
    model = ProjectModelV1(name="test_project", version="1.0.0")

    generated_recipe = GeneratedRecipe.from_model(model, Path("."))

    generated_recipe.recipe.package.name = "new_package_name"
    assert str(generated_recipe.recipe.package.name) == "new_package_name"


def test_package_dependency_modification() -> None:
    """Test initialization of ProjectModelV1."""
    model = ProjectModelV1(name="test_project", version="1.0.0")

    generated_recipe = GeneratedRecipe.from_model(model, Path("."))

    generated_recipe.recipe.requirements.build.append(ItemPackageDependency("test_package"))
    generated_recipe.recipe.requirements.build.append(ItemPackageDependency("test_package"))
    
    assert len(generated_recipe.recipe.requirements.build) == 2



def test_conditional_item() -> None:
    """Test setting of conditional item."""
    
    conditional = ConditionalPackageDependency("os == 'linux'", ListOrItemPackageDependency(["package1"]), ListOrItemPackageDependency(["package3"]))
    item = ItemPackageDependency.new_from_conditional(conditional)

    item.conditional.condition = "jora"

    # this is a known issue with the current implementation
    assert item.conditional.condition == "os == 'linux'"
