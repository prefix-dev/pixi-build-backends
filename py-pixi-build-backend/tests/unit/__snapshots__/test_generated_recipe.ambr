# serializer version: 1
# name: test_generated_recipe_from_model
  '''
  context: {}
  package:
    name: test_project
    version: 1.0.0
  source:
  - path: .
    sha256: null
  build:
    number: null
    script:
      content: []
      env: {}
      secrets: []
  requirements:
    build: []
    host: []
    run: []
    run_constraints: []
  tests: []
  about: null
  extra: null
  
  '''
# ---
